### PR TITLE
ui/datasets: Refactor selectors

### PR DESCRIPTION
**Pull request by ssangervasi**
### Code Changes

- Track dataset by fides key instead of object copy and nulls
  - Previously we've stored a copy of the dataset returned by the get-by-key query,
  but only every use its key property. With this change we only store the key and
  use the object from the query directly.
  - As part of this, I also converted a lot of `null` to `undefined`. There different
  opinions on this, but undefined is generally more useful in a TS codebase because:

    1. It plays nicely with optional function arguments.
    2. It can be represented by a single question mark in interfaces. (See the State
       changes in this commit.)
    3. It doesn't have null's `typeof null === "object"` confusion.

- Single state for edit drawer
  - This resolves some naming inconsistencies and also makes it explicit that only one drawer can ever be open.
- Extract collection lookups to selector
- Extract field lookups to selector
- Extract field types into Cell component

### Steps to Confirm

* [ ] The dataset editing experience should be unchanged.

### Pre-Merge Checklist

* [x] All CI Pipelines Succeeded
* Documentation Updated:
  - [ ] documentation complete, or draft/outline provided (tag docs-team to complete/review on this branch)
  - [ ] documentation issue created (tag docs-team to complete issue separately)
* [ ] Issue Requirements are Met
* [x] Relevant Follow-Up Issues Created
* [ ] Update `CHANGELOG.md`

### Description Of Changes

These changes set me up for #1075. Layering in classification information means a lot of nested lookups with fallbacks. Well-organized selectors are going to make this possible.

### CODE REVIEW

File: clients/ctl/admin-ui/src/features/dataset/dataset.slice.ts [hunk 1]
```diff
@@ -1,21 +1,21 @@
...
 
 import type { AppState } from "~/app/store";
 import { Dataset, GenerateRequestPayload, GenerateResponse } from "~/types/api";
 
+import { EditableType } from "./types";
+
 export interface State {
-  activeDataset: Dataset | null;
+  activeDatasetFidesKey?: string;
   // collections and fields don't have unique IDs, so we have to use their index
```
Comment (ssangervasi) [inline clients/ctl/admin-ui/src/features/dataset/dataset.slice.ts:11]: We may want another refactor to change this index approach. The classification's fields [are going to be](https://github.com/ethyca/fidesctl-plus/pull/106/files#r974617237) identified by the field name, so we'll already have one place we identify fields by name. Also, I feel like I've seen a discussion somewhere to the effect of "if a field changes name, it's a new field as far as fides is concerned" but I can't recall the specifics.

Comment (allisonking) [reply clients/ctl/admin-ui/src/features/dataset/dataset.slice.ts:11]: yah this approach was \*shudder\*. the danger here is that we don't enforce the names to be unique. there's an open ticket for this in fideslang: https://github.com/ethyca/fides/pull/718#issuecomment-1148046497

**Comment by allisonking** [comment]
heh @LKCSmith and I had this discussion around `null` vs `undefined` [a while ago](https://github.com/ethyca/fides/pull/716#issuecomment-1143759657) and settled on `null` over `undefined`, hence why there's so many `null`s around. I don't really have a strong opinion either way and am happy to go back to undefined, but thought I'd surface that old discussion since it might be relevant

**Comment by LKCSmith** [comment]
> heh @LKCSmith and I had this discussion around `null` vs `undefined` [a while ago](https://github.com/ethyca/fides/pull/716#issuecomment-1143759657) and settled on `null` over `undefined`, hence why there's so many `null`s around. I don't really have a strong opinion either way and am happy to go back to undefined, but thought I'd surface that old discussion since it might be relevant

I'm wondering if this is just a practice I carried over from previous jobs that has since changed or evolved in some way. I think I remember questioning if it was still a best practice or not,  but the tradeoffs here sound like they would simplify the TS a bit. I think I generally don't like how it looks with setting or dispatching undefined over null -- it just seems weird to me in a Redux context.
But, if this makes our codebase easier to read or follow though, I'm amenable to it

**Review by allisonking — APPROVED** [review]
nice refactor! looks like a test is failing though

**Comment by ssangervasi** [comment]
> had this discussion around `null` vs `undefined` a while ago

Classic! it's one of those JS classic things that is unreasonably confusing and gets plenty of back-and-forth in a code base (or in random online threads).

>I'm wondering if this is just a practice I carried over from previous jobs that has since changed or evolved in some way.

Yeah, there was a bit of an argument in favor of `null` when we didn't have TS to make a distinction between "this is a value, but nothing is there at the moment" and "this ain't a value at all". I think that's the logic of some built-in APIs like `document.getElementById` which returns null when there is no match. But then JS steps on its own feet with `array.find` returning `undefined` when there is no match!

In TS the benefit of `optional?:` working with undefined in function params and interfaces sways me away from null. 
Though even in plain JS, question mark chaining falls back to undefined, so something like `state.data?.maybe?.value` could result in multiple "empty" values if you use null, but only one if you consistently use undefined.

**Review by LKCSmith — APPROVED** [review]
Thanks for explaining the evolution of the null / undefined discussion! 😆

**Pull request merged**